zaza/openstack/charm_tests/vault/utils.py

```diff
@@ -117,30 +118,26 @@ def get_clients(units=None, cacert=None):
     return clients
 
 
+@tenacity.retry(
+    retry=tenacity.retry_if_exception_type((
+        ConnectionRefusedError,
+        urllib3.exceptions.NewConnectionError,
+        urllib3.exceptions.MaxRetryError,
+        requests.exceptions.ConnectionError)),
+    reraise=True,  # if all retries failed, reraise the last exception
```

Comment:
If all retries failed, it will now raise the last exception, other than Exception("Cannot connect"), which slightly changed the original behavior. Since Exception is base class, it shouldn't affect the upstream code, in theory.